experiments/tf_trainer/common/model_trainer.py

```diff
@@ -233,30 +239,136 @@ def _add_estimator_key(self, estimator, example_key_name):
+    
+    event_file = os.path.join(eval_event_dir, event_files[-1])
+
+    # Use the best step to find the best checkpoint.
+    best_step = self._get_best_step_from_event_file(event_file, metrics_key,
+      is_first_metric_better_fn)
+    
+    # If we couldn't find metrics_key in the event file, try again using loss.
+    if best_step is None:
+      print("Metrics key %s not found in metrics, using 'loss' as metric key." %
+            metrics_key)
+      metrics_key = "loss"
+      # Want the checkpoint with the lowest loss
+      is_first_metric_better_fn = lambda x, y: x < y
+
+      best_step = self._get_best_step_from_event_file(event_file, metrics_key,
+        is_first_metric_better_fn)
+
+    if best_step is None:
+      raise ValueError("Couldn't find 'loss' metric in event file %s." % event_file)
+
+    best_checkpoint_path = None
+    for checkpoint_path in checkpoints:
+      version = int(checkpoint_path.split('-')[-1])
+      if version == best_step:
+        best_checkpoint_path = checkpoint_path
+
+    if not best_checkpoint_path:
+      raise ValueError("Couldn't find checkpoint for best_step = %d." % best_step)
+
+    return best_checkpoint_path
+
+
+  def _get_list_checkpoint(self,
+    n_export,
+    model_dir,
+    metrics_key,
+    is_first_metric_better_fn):
     """Get the checkpoints that we want to export.
 
     Args:
       n_export: Number of models to export.
       model_dir: Directory containing the checkpoints.
+      metrics_key: The metric by which to determine the best checkpoint to save.
+      is_first_metric_better_fn: Comparison function to find best metric. Takes
+          in as arguments two numbers, returns true if first is better than
+          second. Default function says larger is better. Default value works for
+          AUC: higher is better.
 
     Returns:
-      List of checkpoint path.
+      List of checkpoint paths to export.
 
     If n_export==1, we take only the last checkpoint.
+    If n_expost==-1, we take the best checkpoint, according to `metrics_key` and
+      `is_first_metric_better_fn`. The remaining checkpoints are deleted.
     Otherwise, we consider the list of steps for each for which we have a
-    checkpoint.
-    Then we choose n_export number of checkpoints such as their steps are as
-    equidistant as possible.
+    checkpoint. Then we choose n_export number of checkpoints such that their
+    steps are as equidistant as possible.
     """
-
     checkpoints = file_io.get_matching_files(
         os.path.join(model_dir, 'model.ckpt-*.index'))
     checkpoints = [x.replace('.index', '') for x in checkpoints]
     checkpoints = sorted(checkpoints, key=lambda x: int(x.split('-')[-1]))
 
     if n_export == 1:
       return [checkpoints[-1]]
+    elif n_export == -1:
+      # If we want only the best checkpoint, delete the others.
+      best_ckpt = self._get_best_checkpoint(checkpoints, metrics_key,
+        is_first_metric_better_fn)
```

Comment:
nit: I believe that the indentation is not correct.
is_first_metric_fn should be aligned with checkpoints?